File: Python-vs-R-for-Machine-Learning -version-1/code/r/run_security_tests.R
# 2018-compatible, Windows/POSIX safe
# R security tests analogous to the Python suite.
# T1 print(), T2 summary(), T3 indexing, T4 allowlist, T5 CSV control, T6 integrity check.

# --- make working dir = script folder ---
args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
if (length(file_arg)) {
  script_path <- normalizePath(sub("^--file=", "", file_arg))
  script_dir  <- dirname(script_path)
  setwd(script_dir)
}

ART <- "artifacts"
dir.create(ART, showWarnings = FALSE, recursive = TRUE)

# Cross-platform echo helper with safe quoting
echo_msg <- function(msg) {
  if (.Platform$OS.type == "windows") {
    system(paste("cmd /c echo", msg), invisible = TRUE)
  } else {
    # Use shQuote to avoid shell-quoting pitfalls
    system(paste("sh -c", shQuote(paste("echo", msg))), invisible = TRUE)
  }
}

# Malicious S3 methods for class RCE (benign demo payloads)
print.RCE <- function(x, ...) {
  echo_msg("*** malicious RDS executed via print (demo) ***")
  invisible(0L)
}
summary.RCE <- function(object, ...) {
  echo_msg("*** malicious RDS executed via summary (demo) ***")
  invisible(0L)
}
`[.RCE` <- function(x, ...) {
  echo_msg("*** malicious RDS executed via indexing (demo) ***")
  NextMethod("[")
}

# Load allowlist helper
source("safe_loader.R")

ok <- function(label) cat("[OK] ", label, "\n", sep = "")
fail <- function(label, e) { cat("[FAIL] ", label, ": ", conditionMessage(e), "\n", sep = ""); stop(e) }

# Utility: write baseline malicious.rds
write_malicious <- function() {
  rce_obj <- structure(list(msg = "*** malicious RDS executed (demo) ***"), class = "RCE")
  saveRDS(rce_obj, file = file.path(ART, "malicious.rds"))
}

t1_print <- function() {
  label <- "T1 RCE via print() on RCE object"
  tryCatch({
    if (!file.exists(file.path(ART, "malicious.rds"))) write_malicious()
    obj <- readRDS(file.path(ART, "malicious.rds"))
    res <- print(obj)  # triggers print.RCE
    if (!is.integer(res) && !identical(res, 0L)) stop("print() did not return invisible 0L")
    ok(label)
  }, error = function(e) fail(label, e))
}

t2_summary <- function() {
  label <- "T2 RCE via summary() on RCE object"
  tryCatch({
    obj <- readRDS(file.path(ART, "malicious.rds"))
    res <- summary(obj)  # triggers summary.RCE
    ok(label)
  }, error = function(e) fail(label, e))
}

t3_indexing <- function() {
  label <- "T3 RCE via indexing method `[.RCE`"
  tryCatch({
    obj <- readRDS(file.path(ART, "malicious.rds"))
    tmp <- obj[1]  # triggers `[.RCE`
    ok(label)
  }, error = function(e) fail(label, e))
}

t4_allowlist_blocks <- function() {
  label <- "T4 Allowlist blocks disallowed class before use"
  tryCatch({
    blocked <- FALSE
    tryCatch({
      safe_read_rds(file.path(ART, "malicious.rds"))
    }, error = function(e) { blocked <<- TRUE })
    if (!blocked) stop("safe_read_rds did not block RCE class")
    ok(label)
  }, error = function(e) fail(label, e))
}

t5_csv_control <- function() {
  label <- "T5 CSV control: no code execution"
  tryCatch({
    df <- data.frame(x = 1:3, y = c("a","b","c"), stringsAsFactors = FALSE)
    csv_path <- file.path(ART, "control.csv")
    write.csv(df, csv_path, row.names = FALSE)
    df2 <- read.csv(csv_path, stringsAsFactors = FALSE)
    if (!identical(df, df2)) stop("CSV roundtrip mismatch")
    ok(label)
  }, error = function(e) fail(label, e))
}

t6_integrity_check <- function() {
  label <- "T6 Integrity check (MD5 detects tamper)"
  tryCatch({
    path <- file.path(ART, "malicious.rds")
    if (!file.exists(path)) write_malicious()
    md5_known <- tools::md5sum(path)[[1]]
    # Tamper: flip one byte in a copy
    tampered <- file.path(ART, "malicious_tampered.rds")
    bytes <- readBin(path, what = "raw", n = file.info(path)$size)
    i <- min(10L, length(bytes))
    if (i < 1L) stop("file too small to tamper")
    bytes[i] <- as.raw(bitwXor(as.integer(bytes[i]), 0xFF))
    writeBin(bytes, tampered)
    md5_tampered <- tools::md5sum(tampered)[[1]]
    if (identical(md5_tampered, md5_known)) stop("Tamper not detected by MD5")
    ok(label)
  }, error = function(e) fail(label, e))
}

cat("Running R serialization security tests (T1–T6)...\n\n")
t1_print()
t2_summary()
t3_indexing()
t4_allowlist_blocks()
t5_csv_control()
t6_integrity_check()
cat("\nAll tests completed.\n")

File: Python-vs-R-for-Machine-Learning -version-1/code/py/run_security_tests.py
# 2018-compatible test runner (targets Python 3.6 era).
import os
import sys
import subprocess
import hashlib

ROOT = os.path.dirname(os.path.abspath(__file__))
ART = os.path.join(ROOT, "artifacts")
GEN = os.path.join(ROOT, "generate_malicious_pickle.py")
LOAD = os.path.join(ROOT, "deserialization_test.py")

def run(cmd, cwd=None):
    p = subprocess.Popen(cmd, cwd=cwd, stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT, universal_newlines=True)
    out, _ = p.communicate()
    return p.returncode, out

def assert_in(text, haystack, msg):
    if text not in haystack:
        raise AssertionError(msg + "\nExpected substring: {t}\nOutput:\n{h}".format(t=text, h=haystack))

def print_ok(label):
    sys.stdout.write("[OK] {l}\n".format(l=label))

def print_fail(label, e):
    sys.stdout.write("[FAIL] {l}: {e}\n".format(l=label, e=e))
    raise

def ensure_artifacts_dir():
    if not os.path.isdir(ART):
        os.makedirs(ART)

def t1_baseline_reduce():
    label = "T1 Baseline RCE via __reduce__ (protocol=2)"
    try:
        code, out = run([sys.executable, GEN, "--protocol", "2", "--method", "reduce"])
        if code != 0:
            raise RuntimeError(out)
        code, out = run([sys.executable, LOAD])
        if code != 0:
            raise RuntimeError(out)
        # Looser substring so it works across shells
        assert_in("malicious pickle executed (demo)", out, "Echo marker not printed")
        assert_in("Loaded object type: <class 'int'>", out, "Return type not int")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t2_reduce_ex():
    label = "T2 RCE via __reduce_ex__ (protocols 0-3)"
    try:
        for proto in ("0", "1", "2", "3"):
            outpath = os.path.join(ART, "malicious_p{p}.pkl".format(p=proto))
            code, out = run([sys.executable, GEN, "--protocol", proto, "--method", "reduce_ex", "--out", outpath])
            if code != 0:
                raise RuntimeError(out)
            code, out = run([sys.executable, LOAD, "--path", outpath])
            if code != 0:
                raise RuntimeError(out)
            assert_in("malicious pickle executed via __reduce_ex__ (demo)", out,
                      "Missing __reduce_ex__ marker for protocol " + proto)
            assert_in("Loaded object type: <class 'int'>", out, "Return type not int")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t3_protocol_matrix():
    label = "T3 Protocol matrix (__reduce__, protocols 0-3)"
    try:
        for proto in ("0", "1", "2", "3"):
            outpath = os.path.join(ART, "malicious_reduce_p{p}.pkl".format(p=proto))
            code, out = run([sys.executable, GEN, "--protocol", proto, "--method", "reduce", "--out", outpath])
            if code != 0:
                raise RuntimeError(out)
            code, out = run([sys.executable, LOAD, "--path", outpath])
            if code != 0:
                raise RuntimeError(out)
            assert_in("malicious pickle executed (demo)", out, "Missing marker for protocol " + proto)
            assert_in("Loaded object type: <class 'int'>", out, "Return type not int")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t4_allowlist_unpickler_blocks():
    label = "T4 Allowlist Unpickler blocks os.system payload"
    try:
        baseline = os.path.join(ART, "malicious.pkl")
        if not os.path.exists(baseline):
            code, out = run([sys.executable, GEN, "--protocol", "2", "--method", "reduce"])
            if code != 0:
                raise RuntimeError(out)

        # Import safe_unpickler and try to load: expect failure
        code, out = run([sys.executable, "-c",
                         "from safe_unpickler import safe_load; "
                         "f=open(r'%s','rb'); "
                         "safe_load(f)" % baseline],
                        cwd=ROOT)
        if code == 0:
            raise AssertionError("Safe unpickler unexpectedly loaded the payload")
        if "blocked:" not in out and "UnpicklingError" not in out:
            raise AssertionError("Expected UnpicklingError with 'blocked:' in output")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t5_json_control():
    label = "T5 JSON control (no code execution)"
    try:
        py = (
            "import json; "
            "payload={'x':1,'y':[2,3]}; "
            "s=json.dumps(payload); "
            "print(json.loads(s))"
        )
        code, out = run([sys.executable, "-c", py])
        if code != 0:
            raise RuntimeError(out)
        assert_in("{'x': 1, 'y': [2, 3]}", out, "JSON round-trip failed")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t6_integrity_check():
    label = "T6 Integrity check (SHA256, detects tamper)"
    try:
        baseline = os.path.join(ART, "malicious.pkl")
        if not os.path.exists(baseline):
            code, out = run([sys.executable, GEN, "--protocol", "2", "--method", "reduce"])
            if code != 0:
                raise RuntimeError(out)
        data = open(baseline, "rb").read()
        h_known = hashlib.sha256(data).hexdigest()
        tampered = bytearray(data)
        idx = min(10, max(0, len(tampered) - 1))
        tampered[idx] ^= 0xFF
        h_tampered = hashlib.sha256(tampered).hexdigest()
        if h_tampered == h_known:
            raise AssertionError("Tamper not detected by hash")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def t7_constrained_execution():
    label = "T7 Constrained execution (subprocess demo)"
    try:
        code, out = run([sys.executable, GEN, "--protocol", "2", "--method", "reduce"])
        if code != 0:
            raise RuntimeError(out)
        code, out = run([sys.executable, LOAD])
        if code != 0:
            raise RuntimeError(out)
        assert_in("malicious pickle executed (demo)", out, "Marker not printed")
        print_ok(label)
    except Exception as e:
        print_fail(label, e)

def main():
    sys.stdout.write("Running Python serialization security tests (T1–T7)...\n\n")
    ensure_artifacts_dir()
    t1_baseline_reduce()
    t2_reduce_ex()
    t3_protocol_matrix()
    t4_allowlist_unpickler_blocks()
    t5_json_control()
    t6_integrity_check()
    t7_constrained_execution()
    sys.stdout.write("\nAll tests completed.\n")

if __name__ == "__main__":
    main()
